### ISSUE TITLE
Styling isn't always refreshed far enough

### ISSUE THREAD

**Issue by kirbyfan64** [howl-editor/howl#185]
Repro:

1. Create a new C++ source file, `x.cpp`.

2. Enter the following, preceded by one blank line:

  ```c++
  Y {}
  ```

3. In the blank line, type in `class`. Note that `Y` *should* be highlighted as a `type_def`, but it isn't.

4. Press Enter. The word `X` should now be highlighted as a type name. This is correct behavior based on the pattern in the C/C++ lexer. Closing and re-opening Howl would also fix this.

4. After the word `class`, put ` X` (note the space!), so that the line will now say `class X`. Even though `X` is supposed to be highlighted as `type_def` now, `Y` is still highlighted! Closing Howl and re-opening it fixes the problem.

My guess is that `ensure_styled_to` may NOT necessarily go far enough in refreshing the styling.

**Comment by nilnor** [comment]
Well, this is more of a general problem. Typically while editing we want to refresh only the minimum needed, which is typically only the current line. In some case we need to refresh more, such as in the case of open strings, blocks, etc. But in these cases we know this because the refresh of the current line resulted in lexing information indicating this (i.e. eol is styled, with something other than whitespace). In this example we don't get that, since we'll end up with one normally styled keyword. We could extend the default behaviour to refresh a slightly bigger area, which would help with this specific case, but it wouldn't be hard to come up with a new case not working with new the limits. Another route would be to have the lexer provide this additional information in some good way, but as of now I'm unsure how this would look. Or perhaps one could special case these cases directly in the lexer, so that the resulting syntactical information contains this extra information.

**Comment by shalabhc** [comment]
Could this be resolved at the lexer level if it uses a non-standard whitespace capture? E.g. 

```moonscript
iws =  c 'inline-whitespace', space
```

Then using the above instead of `ws` in the classdef rule between 'class' and the name?

Alternatively, since 'whitespace' is only used within other rules, why not make it re-lex if the end of line changed from unstyled to 'whitespace'?

**Comment by nilnor** [comment]
Yes, last I thought about this I thought about having the lexer specifically provide "open" lexing constructs, i.e. prematurely closed constructs it knows should be finished, e.g. "class ". I think this is the right solution for this, as it's very much language dependent and not possible to handle automatically. 

Re the last suggestion, note that both unstyled and 'whitespace' are considered the same from the perspective of the lexer when it determines whether it needs to lex further, and 'whitespace' is perfectly fine in the first place. Even if anything else would do, it makes sense to standardize on a special style for this purpose.

So in short, lexers should do this by having explicit rules for prematurely ended inputs, but this is easiest done by matching e.g. 'class', 'whitespace', `end-of-input`.

**Comment by nilnor** [comment]
As per the previous discussion it's not possible for the styler to know exactly how far the styling must go unless the lexer provides this information in some way. So it's not possible to solve in a generic fashion - instead it's up to the respective lexers to cover these cases (an example of this can be done is in the closing commit fe3a541).
